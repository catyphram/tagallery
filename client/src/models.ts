export enum LIST_MODE {
    MODE_VIEW,
    MODE_VERIFY,
}

export interface Category {
    name: string;
    key: string;
    description?: string;
}

export interface Image {
  file: string;
  assignedCategories?: string[];
  proposedCategories?: string[];
  starredCategory?: string;
}

export interface Config {
  api: string;
}
