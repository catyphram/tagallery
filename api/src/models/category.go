package models

type Category struct {
	Name string `json:"name"`
	Key string `json:"key"`
	Description string `json:"description"`
}

var CategoryFixtures = []*Category {
	{Name: "Test 1", Key: "test1", Description: "Test 1 Category"},
	{Name: "Test 2", Key: "test2", Description: "Test 2 Category"},
}
