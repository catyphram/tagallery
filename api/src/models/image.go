package models

type Image struct {
	Name string `json:"name"`
	AssignedCategories []*Category `json:"assignedCategories"`
	ProposedCategories []*Category `json:"proposedCategories"`
	StarredCategory *Category `json:"starredCategory"`
}

var ImageFixtures = []*Image {
	{Name: "Image 1", AssignedCategories: []*Category{CategoryFixtures[0]}, ProposedCategories: []*Category{}, StarredCategory: nil},
	{Name: "Image 2", AssignedCategories: []*Category{}, ProposedCategories: []*Category{CategoryFixtures[0]}, StarredCategory: nil},
	{Name: "Image 3", AssignedCategories: []*Category{}, ProposedCategories: []*Category{CategoryFixtures[1]}, StarredCategory: nil},
	{Name: "Image 4", AssignedCategories: []*Category{CategoryFixtures[0], CategoryFixtures[1]}, ProposedCategories: []*Category{}, StarredCategory: nil},
	{Name: "Image 5", AssignedCategories: []*Category{}, ProposedCategories: []*Category{}, StarredCategory: CategoryFixtures[1]},
}
