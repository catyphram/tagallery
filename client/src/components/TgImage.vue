<template>
	<div class="tg-image">
		<div :style="`background-image: url(${image.file})`" />
	</div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator';
import { Image } from '@/models';

@Component
export default class TgMenu extends Vue {
  @Prop({default: {}})
  public image!: Image;
}
</script>

<style lang="scss">
	.tg-image {
		position: relative;
		width: calc(100% - 10px);
		margin: 5px;
		&:after {
			content: "";
			display: block;
			padding-bottom: 100%;
		}
		> div {
			position: absolute;
			width: 100%;
			height: 100%;
      background-size: cover;
		}
	}
	@media screen and (min-width: 400px) {	.tg-image {	width: calc(100%/2 - 10px);	}	}
	@media screen and (min-width: 600px) {	.tg-image {	width: calc(100%/3 - 10px);	}	}
	@media screen and (min-width: 850px) {	.tg-image {	width: calc(100%/4 - 10px);	}	}
	@media screen and (min-width: 1100px) {	.tg-image {	width: calc(100%/5 - 10px);	}	}
	@media screen and (min-width: 1400px) {	.tg-image {	width: calc(100%/6 - 10px);	}	}
	@media screen and (min-width: 1700px) {	.tg-image {	width: calc(100%/7 - 10px);	}	}
</style>
