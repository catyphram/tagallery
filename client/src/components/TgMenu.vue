<template>
  <div class="tg-menu">
    <form>
      <md-list>
        <md-list-item class="mode-toggle">
          <md-switch v-model="verifyMode" v-on:change="updateMode">Verify Mode</md-switch>
        </md-list-item>
        <template v-if="!categoriesLoading && categories.length > 0">
          <md-list-item v-for="category in categories" :key="category.key">
            <md-button
              v-on:click="$store.dispatch('toggleCategory', category)"
              :class="{'md-raised': $store.getters.isCategorySelected(category)}"
            >{{category.name}}</md-button>
          </md-list-item>
        </template>
        <md-list-item>
          <md-button
            v-on:click="$store.dispatch('toggleListUncategorized')"
            :class="{'md-raised': $store.state.listUncategorized}"
          >Uncategorized</md-button>
        </md-list-item>
      </md-list>
    </form>
  </div>
</template>

<script lang="ts">
import { Component, Watch, Vue } from 'vue-property-decorator';

import { LIST_MODE, Image, Category } from '../models';

@Component
export default class TgMenu extends Vue {
  get verifyMode(): boolean {
    return this.$store.state.listMode === LIST_MODE.MODE_VERIFY;
  }
  get categories(): Category[] {
    return this.$store.state.categories.data;
  }
  get categoriesLoading(): boolean {
    return this.$store.state.categories.loading;
  }
  public updateMode(value: boolean) {
    this.$store.dispatch('setMode', value ? LIST_MODE.MODE_VERIFY : LIST_MODE.MODE_VIEW);
  }
}
</script>

<style scoped lang="scss">
  .md-button {
    display: block;
  }
  .mode-toggle {
    margin-left: 16px;
  }
</style>
