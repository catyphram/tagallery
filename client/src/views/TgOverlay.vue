<template>
	<div class="tg-overlay" v-if="image">
    <div class="tg-overlay__menu">
      <span v-if="categories" class="tg-overlay__categories">
        <md-button
          v-for="categoryItem in categories" :key="categoryItem.category.key"
          @click="() => toggleImageCategory(categoryItem.category)"
          :class="{
            'md-raised': categoryItem.assigned,
            'tg-overlay__category--proposed': categoryItem.proposed,
          }"
        >{{categoryItem.category.name}}</md-button>
      </span>
      <md-icon @click.native="close" class="md-size-2x tg-overlay__close">close</md-icon>
    </div>
    <md-icon @click.native="prevImage" class="md-size-2x">arrow_back</md-icon>
		<img :src="image.file" alt="image" />
    <md-icon @click.native="nextImage" class="md-size-2x">arrow_forward</md-icon>
	</div>
</template>

<script lang="ts">
import { Component, Watch, Vue } from 'vue-property-decorator';

import { Image, Category } from '../models';

@Component
export default class TgOverlay extends Vue {
  get image(): Image {
    return this.$store.state.selectedImage !== undefined &&
      this.$store.state.images.data[this.$store.state.selectedImage] ?
      this.$store.state.images.data[this.$store.state.selectedImage] :
      undefined;
  }
  get categories(): Array<{
    category: Category,
    assigned: boolean,
    proposed: boolean,
  }> | undefined {
    if (this.image !== undefined) {
      return this.$store.state.categories.data.map((category: Category) => {
        return {
          category,
          assigned: this.imageHasAssignedCategory(category),
          proposed: this.imageHasProposedCategory(category),
        };
      });
    }
  }
  public prevImage() {
    if (this.$store.state.selectedImage) {
      this.$store.commit('selectImage', { index: this.$store.state.selectedImage - 1 });
    }
  }
  public nextImage() {
    if (this.$store.state.selectedImage !== undefined) {
      this.$store.commit('selectImage', { index: this.$store.state.selectedImage + 1 });
    }
  }
  public close() {
    this.$store.commit('selectImage', { index: undefined });
  }
  public toggleImageCategory(category) {
    this.$store.dispatch('toggleImageCategory', { image: this.image, category });
  }
  protected imageHasAssignedCategory(category: Category) {
    return !!(this.image.assignedCategories && this.image.assignedCategories.find((assignedCategory) => {
      return assignedCategory === category.key;
    }));
  }
  protected imageHasProposedCategory(category: Category) {
    return !!(this.image.proposedCategories && this.image.proposedCategories.find((proposedCategory) => {
      return proposedCategory === category.key;
    }));
  }
}
</script>

<style lang="scss">
	.tg-overlay {
    background: #424242;
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    right: 0;
    z-index: 100;
    display: flex;
    justify-content: center;
    align-items: center;

    &__menu {
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      display: inline-flex;
    }

    &__categories {
      flex: 1;
      display: inline-flex;
      justify-content: center;
      margin-left: 48px;
    }
    &__category {
      &--proposed {
        color: deepskyblue !important;
      }
    }

    &__close {
      flex: 1;
      display: inline-flex;
      justify-content: flex-end;
      flex-grow: 0;
    }
	}
</style>
