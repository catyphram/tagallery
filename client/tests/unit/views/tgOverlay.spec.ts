import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import { MdIcon } from 'vue-material/dist/components';
import TgOverlay from '@/views/TgOverlay.vue';
import { Image } from '@/models';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(MdIcon);

const state: {
  images: {
    data: Image[],
  },
  selectedImage?: number,
} = {
  images: {
    data: [
      { file: 'test1.jpg' }, { file: 'test2.jpg' }, { file: 'test3.jpg' },
    ] as Image[],
  },
};

const mutations = {
  selectImage(state, { index }: { index?: number }) {
    state.selectedImage = index;
  },
};

const store = new Vuex.Store({
  mutations,
  state,
});

describe('TgOverlay.vue', () => {
  it('image should return undefined when no image is selected', () => {
    const wrapper = shallowMount(TgOverlay, { store, localVue });
    const view = wrapper.vm as any;
    expect(view.image).toBeUndefined();
  });

  it('image should select the selected image from the store', () => {
    const wrapper = shallowMount(TgOverlay, { store: new Vuex.Store({
      state: {
        ...state,
        selectedImage: 0,
      },
    }), localVue });
    const view = wrapper.vm as any;
    expect(view.image).toEqual(state.images.data[0].file);
  });

  it('prevImage should not select the previous image if no image is selected', () => {
    const wrapper = shallowMount(TgOverlay, { store, localVue });
    const view = wrapper.vm as any;

    store.commit = jest.fn();
    view.prevImage();

    expect(store.commit).not.toHaveBeenCalled();
  });

  it('prevImage should select the next image', () => {
    const localStore = new Vuex.Store({
      state: {
        ...state,
        selectedImage: 1,
      },
      mutations,
    });
    const wrapper = shallowMount(TgOverlay, { store: localStore, localVue });
    const view = wrapper.vm as any;

    localStore.commit = jest.fn();
    view.prevImage();

    expect(localStore.commit).toHaveBeenCalledWith('selectImage', { index: 0 });
  });

  it('nextImage should not select the next image if no image is selected', () => {
    const wrapper = shallowMount(TgOverlay, { store, localVue });
    const view = wrapper.vm as any;

    store.commit = jest.fn();
    view.nextImage();

    expect(store.commit).not.toHaveBeenCalled();
  });

  it('nextImage should select the next image', () => {
    const localStore = new Vuex.Store({
      state: {
        ...state,
        selectedImage: 1,
      },
      mutations,
    });
    const wrapper = shallowMount(TgOverlay, { store: localStore, localVue });
    const view = wrapper.vm as any;

    localStore.commit = jest.fn();
    view.nextImage();

    expect(localStore.commit).toHaveBeenCalledWith('selectImage', { index: 2 });
  });

  it('close should unselect the image', () => {
    const localStore = new Vuex.Store({
      state: {
        ...state,
        selectedImage: 0,
      },
      mutations,
    });
    const wrapper = shallowMount(TgOverlay, { store: localStore, localVue });
    const view = wrapper.vm as any;

    localStore.commit = jest.fn();
    view.close();

    expect(localStore.commit).toHaveBeenCalledWith('selectImage', { index: undefined });
  });
});
