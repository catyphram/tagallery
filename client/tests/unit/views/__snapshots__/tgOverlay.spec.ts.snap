// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TgOverlay.vue categories should select and map the categories from the store 1`] = `
Array [
  Object {
    "assigned": false,
    "category": Object {
      "key": "category1",
      "name": "Category 1",
    },
    "proposed": true,
  },
  Object {
    "assigned": true,
    "category": Object {
      "key": "category2",
      "name": "Category 2",
    },
    "proposed": false,
  },
]
`;

exports[`TgOverlay.vue toggleImageCategory should toggle a assigned category of an image 1`] = `
Array [
  Array [
    "toggleImageCategory",
    Object {
      "category": Object {
        "key": "category1",
        "name": "Category 1",
      },
      "image": Object {
        "file": "test1.jpg",
      },
    },
  ],
]
`;
